### ISSUE TITLE
Cannot pattern match on Bool literal in a Tuple

### ISSUE THREAD

**Issue by itsgreggreg** [mint-lang/mint#377]
In this sandbox: https://sandbox.mint-lang.com/sandboxes/xi4p_mMixnc-yA I'm trying to pattern match on a tuple that contains a boolean. In the function `matchOnTuple` the `{true, m}` branch will always succeed. It seems that the Bool literals are being seen as variable names in the pattern match by the compiler?

 My guess is this is a bug for both ergonomics and because you couldn't access the value stored in that variable as any attempt to use `true` in an expression would be seen as a Bool literal.

**Comment by gdotdesign** [comment]
> It seems that the Bool literals are being seen as variable names in the pattern match by the compiler?

Yes. 

There are specific cases to allow parsing of full Bool literal tuples https://github.com/mint-lang/mint/blob/master/src/parsers/case_branch.cr#L11 but I didn't think about partial Bool literal tuples.

The main issue is that the tuple destructing has the same syntax as a tuple literal which makes it ambigous:

```
try {
  a = "A"
  b = "B"

  case ({"A", "B"}) {
    {a, b} => true
    => false
  }
}
```

How can we determine if the example above uses a tuple literal to match against or tuple destructuring? The same is true for arrays. I'm curious how for example Elm handles this?

So I'm not sure how to resolve this, probably by changing the syntax of the destructurings, also your example makes it more complex because it mixes the two: `{true, m}` I'm guessing it's a destructuring but only for the second element of the tuple?

**Comment by itsgreggreg** [comment]
Gotcha, what I think I'm understanding is that you cannot partially match a structure in mint? You are either matching on known values _or_ capturing structure, but never a mixture of both?

Per your example above the common answer is that in `{a, b} => true` `a` and `b` are variables created in the match itself that are matching a sub-structure and are not values from a wider scope. 

In Elm, there's no way to match on the values of variables and attempting to do so will throw a variable shadowing error:
https://ellie-app.com/csC7Dz67k9Ja1

In Elixir this is also the default and if you want to match on the value of a variable you must use the pin operator https://elixir-lang.org/getting-started/pattern-matching.html

I believe this is also the case in Rust, though that language has so much syntax I might just not know about their variable pinning operator.



Most languages treat it as a new variable binding and most languages don't have a way to pattern match on the values of varibles.

**Comment by itsgreggreg** [comment]
The part about it being a tuple literal or tuple pattern I think is unimportant as in either case it's the contents of the tuple we're interested in.

**Comment by gdotdesign** [comment]
> Gotcha, what I think I'm understanding is that you cannot partially match a structure in mint? You are either matching on known values _or_ capturing structure, but never a mixture of both?

Currently yes, this is the case.



> In Elixir this is also the default and if you want to match on the value of a variable you must use the pin operator [elixir-lang.org/getting-started/pattern-matching.html](https://elixir-lang.org/getting-started/pattern-matching.html)

This I like since we can use it to differentiate during parsing.

**Comment by itsgreggreg** [comment]
> This I like since we can use it to differentiate during parsing.

Agreed and I find pinning to be very useful in elixir and something I wish I had when I write elm.

**Comment by gdotdesign** [comment]
This will be fixed by #593
